odo/backends/avro.py

```diff
@@ -0,0 +1,252 @@
+from __future__ import absolute_import, division, print_function
+
+import errno
+import os
+import uuid
+
+from avro import schema, datafile, io
+from avro.schema import AvroException
+import pandas as pd
+from datashape import discover, var, Record, Map, Var, \
+    Option, null, string, int32, int64, float64, float32, boolean
+from collections import Iterator
+from ..append import append
+from ..convert import convert
+from ..resource import resource
+from ..temp import Temp
+
+AVRO_TYPE_MAP = {
+    'string': string,
+    'int': int32,
+    'long': int64,
+    'null': null,
+    'double': float64,
+    'float': float32,
+    'bool': boolean,
+    'map': Map,
+    'record': Record,
+    'array': Var,
+}
+
+class AVRO(object):
+    """Wrapper object for reading and writing an Avro container file
+
+    Parameters
+    ----------
+
+    uri : str
+        uri of avro data
+
+    schema : avro.schema.Schema
+        User specified Avro schema object.  Used to decode file or serialize new records to file.
+         schema is required to create a new Avro file.
+         If reading or appending to an existing Avro file, the writers_schema embedded in that file
+         will be used.
+
+    codec : str
+        compression codec.  Valid values: 'null', 'deflate', 'snappy'
+
+    """
+    def __init__(self, uri, schema=None, codec='null', **kwargs):
+        self._uri = uri
+        self._schema = schema
+        self._codec = codec
+        self._kwargs = kwargs #CURRENTLY UNUSED
+
+        if not schema:
+            sch = self._get_writers_schema()
+            if sch is None:
+                raise AvroException("Couldn't extract writers schema from '{0}'.  User must provide a valid schema".format(uri))
+            self._schema = sch
+
+    def __iter__(self):
+        return self.reader
+
+    def next(self):
+        return self.reader.next()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
+        # Perform a close if there's no exception
+        if type is None:
+            self.reader.close()
+            self.writer.close()
+
+    def _get_writers_schema(self):
+        """
+        Extract writers schema embedded in an existing Avro file.
+        """
+        reader = self.reader
+        return schema.parse(self.reader.meta['avro.schema']) if reader else None
+
+    uri = property(lambda self: self._uri)
```

Comment:
uri, schema, and codec feel like properties that should be immutable.  If a user changed them, you would need to refresh any open readers and writers.  It feels more natural to expect the user to create a new AVRO object rather than reusing an existing one.